substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java

```diff
@@ -0,0 +1,552 @@
+        } else {
+            log("buildType() %s %s size=%d - begin", typeEntry.typeKind().name(), typeEntry.getTypeName(), typeEntry.getSize());
+            switch (typeEntry.typeKind()) {
+                case PRIMITIVE: {
+                    typeRecord = types.getExistingType(typeEntry);
+                    break;
+                }
+                case ARRAY:
+                case ENUM:
+                case INSTANCE:
+                case INTERFACE: {
+                    typeRecord = buildStructureTypeEntry((StructureTypeEntry) typeEntry);
+                    break;
+                }
+                case HEADER: {
+                    /*
+                     * The bits at the beginning of an Object: contains pointer to DynamicHub.
+                     */
+                    assert typeEntry.getTypeName().equals(OBJ_HEADER_NAME);
+                    typeRecord = buildStructureTypeEntry((HeaderTypeEntry) typeEntry);
+                    break;
+                }
+            }
+        }
+        assert typeRecord != null;
+        log("buildType end: %s", typeRecord);
+        return typeRecord;
+    }
+
+    /**
+     * Add type records for function. In the future add local types when they become available.
+     *
+     * @param entry primaryEntry containing entities whose type records must be added
+     * @return type record for this function (may return existing matching record)
+     */
+    CVTypeRecord buildFunction(CompiledMethodEntry entry) {
+        return buildMemberFunction(entry.getClassEntry(), entry.getPrimary().getMethodEntry());
+    }
+
+    static class FieldListBuilder {
+
+        /* Size of the header part of a field list record (record type + size) */
+        private static final int FIELD_LIST_RECORD_OVERHEAD = Short.BYTES + Short.BYTES;
+
+        final List<CVTypeRecord.FieldRecord> fields = new ArrayList<>();
+
+        FieldListBuilder() {
+        }
+
+        void addField(CVTypeRecord.FieldRecord field) {
+            fields.add(field);
+        }
+
+        int getFieldCount() {
+            return fields.size();
+        }
+
+        CVTypeRecord.CVFieldListRecord buildFieldListRecords(CVTypeSectionBuilder builder) {
+
+            /*
+             * The last FieldList must refer back to the one before it, and must contain the first
+             * fields in the class.
+             */
+
+            CVTypeRecord.CVFieldListRecord currentFieldList = new CVTypeRecord.CVFieldListRecord();
+            Deque<CVTypeRecord.CVFieldListRecord> fl = new LinkedList<>();
+            fl.add(currentFieldList);
+
+            /* Build all Field List records in field order (FIFO). */
+            for (CVTypeRecord.FieldRecord fieldRecord : fields) {
+                if (FIELD_LIST_RECORD_OVERHEAD + currentFieldList.getEstimatedSize() + CVUtil.align4(fieldRecord.computeSize()) >= CV_TYPE_RECORD_MAX_SIZE) {
```

Comment:
I'm not sure I understand this condition: the initial value of `CVFieldListRecord.estimatedSize` essentially already takes into account `FIELD_LIST_RECORD_OVERHEAD`, so why add it again here?

Shouldn't we reserve space for `CVIndexRecord` instead? That is, shouldn't the condition be something like this
```suggestion
                if (currentFieldList.getEstimatedSize() + CVUtil.align4(fieldRecord.computeSize()) > CV_TYPE_RECORD_MAX_SIZE - SPACE_FOR_CV_INDEX_RECORD) {
```
where `SPACE_FOR_CV_INDEX_RECORD` is equal to `CVUtil.align4(new CVIndexRecord(0).computeSize())`?

Reply:
Done.  I've rearranged your suggestion and added some comments explaining the intent of the calculation.